src/include/storage/data_table.h

```diff
@@ -94,18 +93,26 @@ class DataTable {
 
    private:
     friend class DataTable;
+
+    /** Indicates that the iterator should advance to the end of the table. */
+    static constexpr int32_t TILL_END = -1;
+
     /**
      * @warning MUST BE CALLED ONLY WHEN CALLER HOLDS LOCK TO THE LIST OF RAW BLOCKS IN THE DATA TABLE
      */
-    SlotIterator(const DataTable *table, std::list<RawBlock *>::const_iterator block, uint32_t offset_in_block)
-        : table_(table), block_(block) {
-      current_slot_ = {block == table->blocks_.end() ? nullptr : *block, offset_in_block};
+    SlotIterator(const DataTable *table, uint32_t block_index, int32_t num_advances, uint32_t offset_in_block)
+        : table_(table), block_index_(block_index), num_advances_(num_advances) {
+      current_slot_ = {block_index >= table_->blocks_.size() ? nullptr : table->blocks_[block_index], offset_in_block};
```

Comment:
Recommend adding an assertion here that effectively asserts that `offset_in_block == 0` when `nullptr` is used as the block pointer.